### PR TITLE
Fix command line arguments with non-ASCII characters on Windows

### LINKED ISSUE #4369: map_extract fails to open absolute path with Unicode
Reported by cheeser on Discord, works with relative paths.
unknown
Not sure what the issue is. @Robyt3 any idea?

### PR DESCRIPTION
**Pull request by heinrich5991**
Builds upon @Robyt3's #4378. Tries to make less changes to the whole codebase by simply fixing up argc and argv at program start. Fixes #4369.

## Checklist

- [ ] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by heinrich5991** [comment]
Untested on Windows, I don't have a Windows machine. Could someone with a Windows machine test this with non-ASCII parameters?

**Comment by heinrich5991** [comment]
This unfortunately lost the refactorings Robyt3 did in the same commit as the rest of the changes.

File: src/base/system.cpp [hunk 1]
```diff
@@ -3542,6 +3542,49 @@ int pid()
 #endif
 }
 
+void cmdline_fix(int *argc, const char ***argv)
+{
+#if defined(CONF_FAMILY_WINDOWS)
+	int wide_argc = 0;
+	WCHAR **wide_argv = CommandLineToArgvW(GetCommandLineW(), &wide_argc);
```
Comment (Robyt3) [inline src/base/system.cpp:3549]: `wide_argv` can be NULL in case of error, just return if that is the case.

File: src/base/system.cpp [hunk 2]
```diff
@@ -3542,6 +3542,49 @@ int pid()
...
+	mem_zero(new_argv[0], total_size);
+
+	int remaining_size = total_size;
+	for(int i = 0; i < wide_argc; i++)
+	{
+		int size = WideCharToMultiByte(CP_UTF8, 0, wide_argv[i], -1, new_argv[i], remaining_size, NULL, NULL);
+		dbg_assert(size != 0, "WideCharToMultiByte failure");
+
+		remaining_size -= size;
+		wide_argv[i + 1] = wide_argv[i] + size;
```
Comment (Robyt3) [inline src/base/system.cpp:3571]: `wide_argv` -> `new_argv`?

File: src/base/system.cpp [hunk 3]
```diff
@@ -3542,6 +3542,49 @@ int pid()
...
+	}
+
+	wide_argv[wide_argc] = 0;
```
Comment (Robyt3) [inline src/base/system.cpp:3574]: `wide_argv` -> `new_argv`?

**Comment by Robyt3** [comment]
I'm not seeing any obvious error messages the windows build on github, just `Error: Process completed with exit code 1.` I supposed you should also add `cmdline_fix` and `cmdline_free` in the test.cpp main.

I'll test whether this still works later.

The refactorings were mostly minor so I can do them separately eventually. This would also increase the chance of this ever being merged by oy on upstream.

File: src/base/system.cpp (hunk 2 above)
Comment (heinrich5991) [reply src/base/system.cpp:3571]: Done, nicely spotted!

File: src/base/system.cpp (hunk 3 above)
Comment (heinrich5991) [reply src/base/system.cpp:3574]: Done.

File: src/base/system.cpp (hunk 1 above)
Comment (Robyt3) [reply src/base/system.cpp:3549]: Okay, actually not because cmdline_free wouldn't know that the buffers should not be freed in that case.

**Comment by Robyt3** [comment]
I seems you forgot to add the change to the new commit. But with the change it works correctly, I tested a few use cases. I don't know why the run on github is failing though.

File: src/base/system.cpp (hunk 1 above)
Comment (Robyt3) [reply src/base/system.cpp:3549]: `dbg_assert(wide_argv != NULL, "CommandLineToArgvW failure");`?

File: src/base/system.cpp [hunk 4]
```diff
@@ -3542,6 +3542,49 @@ int pid()
...
+	*argc = wide_argc;
+	*argv = (const char **)new_argv;
+#endif
+}
+
+void cmdline_free(int argc, const char **argv)
+{
+#if defined(CONF_FAMILY_WINDOWS)
+	free((void *)*argv);
+	free(argv);
```
Comment (Robyt3) [inline src/base/system.cpp:3584]: `free((char **)argv);` so it compiles with warnings-as-errors on Windows.

**Comment by heinrich5991** [comment]
Superseded by #4423.

**Pull request closed without merging**